KEP-1a: Meta KEP Implementation

justaugustus opened the pull request:
This is a provisional meta-KEP, which discusses iterative improvements to the KEP process.

rel:
- https://github.com/kubernetes/community/issues/2565
- SIG Architecture meeting mentioning the need for a clearer KEP process - https://youtu.be/MMJ-zAR_GbI
- SIG Architecture meeting agreeing to move forward with a KEP process improvement effort to be co-led with SIG PM (@justaugustus / @jdumars) - https://youtu.be/fmlXkN4DJy0

/sig pm architecture
/assign @jdumars @calebamiles @idvoretskyi @timothysc 

Signed-off-by: Stephen Augustus <foo@agst.us>

justaugustus commented:
Adding Tim, Aish, and Aaron to weigh in on Release Team implications.

/assign @tpepper @AishSundar @spiffxp 
/sig release

justaugustus commented:
Adding an explicit hold until we gather some feedback:
/hold

bgrant0607 commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,128 @@
+- Define KEP [DACI]
+
+#### Organize
+
```
Have we resolved the numbering issue?

justaugustus replied:
I've added both options to the KEP. Jaice, Caleb, and I can bikeshed once this gets merged as `provisional`.

timothysc commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Organize
+
+- Move KEPs into [k/features]
```
Just to be honest, this document was confusing up until this point.

timothysc replied:
Also, this KEP will likely be dead after the transition... but whatevers.

justaugustus replied:
@timothysc -- were there specific parts that were confusing? The motivation was "create a KEP that speaks about the improvements we need to make to the process, while ensuring (in the interim) that the process doesn't change underneath SIGs who have adopted it in its' current state".

And yep, this KEP will effectively die once we re-implement KEP-1.

timothysc commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,133 @@
+- Create tooling to present:
+  - KEPs, through some easy to use mechanism e.g., https://keps.k8s.io
+  - Project tracking across SIGs
+  - SIGs roadmaps
```
This might be a bit orthogonal imo.  SIGs react on the fly.

justaugustus replied:
_Currently_, SIGs react on the fly. 
Ideally, moving forward, SIGs should be able to plan and provide some level of soft commit prior to the release cycle, and that information should be aggregatable through the metadata in each KEP. The shape of how exactly to aggregate the data hasn't been defined yet, though.

timothysc replied:
Without specifics, I would remove.

justaugustus replied:
@timothysc -- removed

erictune commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Organize
+
+- Move KEPs into [k/features]
```
Can you say more about how you expect end users to use the new repository structure?

If I am an end user who wants to find out about the design of Kubernetes, what URL should I start at?  For Python, there is one page, called https://www.python.org/dev/peps/ that lists all their PEPs (~=KEPs).

justaugustus replied:
End users should be able to hit https://keps.k8s.io, which would be a redesigned version of https://contributor.kubernetes.io/keps/.

I envision this site / repo having at least three directories:
- `keps/` (KEPs)
- `design-proposals/` (historical design proposals from https://git.k8s.io/community/contributors/design-proposals)
- `arch[itecture]|design/` (design principles of Kubernetes, derived from reorganizing https://git.k8s.io/community/contributors/devel, mentioned [here](https://github.com/kubernetes/community/issues/2565#issuecomment-419185591))

erictune commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,133 @@
+- Enable redirects to [k/keps]
+- Prevent new KEPs and design proposals from landing in [k/community]
+- Correlate existing features with KEPs
+- Fix [KEP numbering races] by adopting one of two schemes:
```
Note that Python solves this, AFAICT, by having a single PEP, called PEP 0, which is an index of all the PEPs. You reserve a number by sending a commit to add a row to the table in PEP 0.

erictune commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,133 @@
+- Correlate existing features with KEPs
+- Fix [KEP numbering races] by adopting one of two schemes:
+  - Issue number of the KEP tracking issue
+  - PR number for the PR that marks the KEP as `accepted`
```
So, then KEPs don't have a number until they are accepted?  This makes talking about them harder?  Also, don't we want to keep a record of non-accepted KEPs as a record of what didn't work?  Python allocates numbers to all proposals and tracks rejected and withdrawn as well as accepted PEPs

justaugustus replied:
Good point about the KEPs that aren't in some accepted state. I've updated the copy to say we'll track based on the GitHub issue number of the KEP tracking issue.

justaugustus commented:
@timothysc @erictune -- PTAL again. Hoping that I've addressed your comments.

I've also added a proposed directory structure as well as the metadata structure that would be included per KEP directory.

spiffxp commented:
ref:https://github.com/kubernetes/test-infra/pull/9437 please document how/why you intend to use the `kind/keps` label

justaugustus commented:
@spiffxp -- I've added notes about the `kind/kep` label, based on what we discussed over here: https://github.com/kubernetes/test-infra/pull/9437#issuecomment-422263046

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+│   ├── feature-01.md
+│   ├── feature-02.md
+│   └── feature-03.md
+├── index.json (required)
```
what is this file?

justaugustus replied:
`index.json` would be gobbled up by Hugo to generate the website.

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+
+#### Organize
+
+- Move KEPs from flat-files to a directory structure:
```
Would KEPs be in per-sig directories as they are now?

justaugustus replied:
KEPs will continue to be in per SIG dirs. This is the layout of a KEP dir within a SIG dir.

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+├── 006-guide-for-teachers.md (required)
+├── 007-guide-for-operators.md (required)
+├── 008-guide-for-project-maintainers.md
+├── 009-graduation-criteria.md (required)
```
Maybe this is covered in the videos, but... this is an awful lot of files.  As a reader, am I going to have to keep jumping between files?  How are existing KEP's going to be migrated to this format? I'm confused that the goals/non-goals aren't required.  Are all of these required when I first submit a KEP, or are some of them required to move a KEP from one status to the next?

spiffxp replied:
If we're deprecating design-proposals, which of these docs is supposed to describe the proposed design/implementation and alternatives considered?

justaugustus replied:
I had the same concern about the about of files. We'll workshop this and get back to you.
It probably makes sense to roll-up the `000..00n` files into one, which includes the currently implemented list:
* [Summary](#summary)
* [Motivation](#motivation)
    * [Goals](#goals)
    * [Non-Goals](#non-goals)
* [Proposal](#proposal)
    * [User Stories [optional]](#user-stories-optional)
      * [Story 1](#story-1)
      * [Story 2](#story-2)
    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
    * [Risks and Mitigations](#risks-and-mitigations)
* [Graduation Criteria](#graduation-criteria)
* [Implementation History](#implementation-history)
* [Drawbacks [optional]](#drawbacks-optional)
* [Alternatives [optional]](#alternatives-optional)

...so maybe no change on that front.

Migration of existing KEPs: haven't nailed that down. Maybe we divide-and-conquer and guide SIGs through the process, similar to what Steering does for SIG Charters?

Goals / Non-goals should be required. Will fix.

> Are all of these required when I first submit a KEP, or are some of them required to move a KEP from one status to the next?

No. We need to define the state machine and entry and exit criteria for each.

> If we're deprecating design-proposals, which of these docs is supposed to describe the proposed design/implementation and alternatives considered?

Should be part of the `Proposal` / `Drawbacks` / `Alternatives` sections, within the KEP.

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+├── experience_reports (required)
+│   ├── alpha-feedback.md
+│   └── beta-feedback.md
+├── features
```
What is the purpose of this directory?

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+├── 008-guide-for-project-maintainers.md
+├── 009-graduation-criteria.md (required)
+├── 010-implementation-history.md
+├── experience_reports (required)
```
What is the purpose of this directory?

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+    - Label incoming KEPs as `kind/kep`
+    - Classify KEP submissions / tracking issues as `kind/kep`, differentiating them from `kind/feature`
+- Move existing design proposals into [k/features]
+- Move existing architectural documents into [k/features]
```
Which ones?

justaugustus replied:
We'd need to nail that down depending on the audience. See @bgrant0607 's comment [here](https://github.com/kubernetes/community/issues/2565#issuecomment-419185591).

I'm wondering if that should be out of scope for this KEP.

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+    - Classify KEP submissions / tracking issues as `kind/kep`, differentiating them from `kind/feature`
+- Move existing design proposals into [k/features]
+- Move existing architectural documents into [k/features]
+- Deprecate design proposals
```
Does this imply from this point forward, everything must be a KEP?

justaugustus replied:
Any significant, multiple-cycle work should be a KEP.
Here's a good example of a design proposal, that is essentially a KEP, that should probably just be in the KEPs dir: https://github.com/kubernetes/community/pull/2514

spiffxp commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,206 @@
+- Create tombstones / redirects to [k/keps]
+- Prevent new KEPs and design proposals from landing in [k/community]
+- Remove `kind/kep` from [k/community] once KEP migration is complete
+- Correlate existing features with KEPs
```
Does this mean turning features into KEPs?

justaugustus replied:
We're introducing the idea of an umbrella term: `Enhancements`.
Enhancements are anything that modifies or adds new functionality to Kubernetes.

So KEPs and Features would both fall under that category.

KEP: multi-release event
Feature: hopefully only a single-release event

Features should roll into a KEP.

I think a good current example might be CoreDNS.
- Implementing CoreDNS == KEP
- CoreDNS as default in kubeadm == Feature

What I meant by this bullet point was that existing Feature issues should be associated to a KEP. We have a line for associating proposals to features in the `k/features` issue description, but not everyone fills it out. We'd want to backfill any missing links here.

justaugustus commented:
@spiffxp -- I made some changes based on your feedback. Thanks!

- Add requirement for glossary of terms
- Add requirement to define KEP state machine, entry / exit criteria
- Include SIG dirs in dir structure (keps/ --> sig-foo/)
- Roll-up extra files in KEP dir
- Describe index.json
- Explain features dir
- Clarify existing features should be linked to KEPs
- Suggest migration of existing KEPs to be shepherded by SIG PM
- AI to figure out moving architectural guides

Spoke w/ @calebamiles and @jdumars and we're thinking `k/enhancements`, instead of `k/keps`.
It captures the spirit of tracking both KEPs and Features, and is less clumsy than looking for a KEP in `https://git.k8s.io/keps/keps`.

erictune commented on keps/0001a-meta-kep-implementation.md:
````diff
@@ -0,0 +1,213 @@
+- Move KEPs from flat-files to a directory structure:
+```
+├── keps # top-level directory
+│   ├── sig-beard # SIG directory
````
Why are KEPs by SIG?  This means people have to look in multiple places to find stuff.  Compare to python PEP-0 where you only have to look at one page to find all the KEPs.

Or the index of all the RFCs from IETF: 
https://tools.ietf.org/rfc/index

jdumars replied:
I feel like repository layout and visual presentation shouldn't be conflated. The job of a KEP is to convey an intent, as well as track how the work to satisfy the intent evolves over time. With that lens, the layout is best oriented to serve the highest stakeholders which are subproject participants first and foremost, SIGs second, consumers like the release team third, and causal observers dead last. Focusing on the current visualization as a starting point feels like a red herring since the KEP process itself, state transition definitions, cli tooling, etc. is not formalized or implemented. Reuse of the OWNERS construct at the subproject level makes a ton of sense.

spiffxp replied:
I'm stuck on repo layout and such because ever since KEPs were first suggested we've been using the "automation will make this more digestible" line, yet the automation hasn't arrived. Until it does, this is the experience we're left with.  Imposing even more friction up front doesn't fill me with confidence that automation will come any sooner.

erictune commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,213 @@
+|   |   |   │   ├── ...
+|   |   |   │   └── feature-n.md
+|   |   |   ├── guides
+|   |   |   |   ├── guide-for-developers.md (required)
```
Guides seem like something that goes on the Docs site.

erictune commented on keps/0001a-meta-kep-implementation.md:
````diff
@@ -0,0 +1,213 @@
+metadata.yaml would contain information that was previously in a KEP's YAML front-matter:
+
+```
+---
````
This strikes me as a lot of metadata required to write a KEP.

calebamiles replied:
should be generated by automation in the general case

jdumars replied:
+1 to automation managing the metadata as much as possible.

erictune commented:
Perhaps this is just me, but I dislike that https://contributor.kubernetes.io/keps/ is organized alphabetically, rather than numerically.  numbers always fit in the left nav bar unlike titles, and the consecutive numbering scheme is largely chronological.  I can't tell if the PR-based numbering scheme gives this desirable property.  It seems like it ties something that belongs to the project (numbering of enhancements) rather too tightly to the specific implementation of PR numbering on github, and the particular github repo in which issue are filed.

calebamiles commented:
@erictune, RE numbering. I believe PR numbers have two very strongly desirable qualities

1. they're monotonically increasing
2. that they're monotonically increasing is somebody else's problem

Now there likely will be "holes" in the sequence of numbers but I think that's a reasonable trade off to avoid the merge conflict hell we've currently got

justaugustus commented:
@erictune -- apologies for not adding that. Just pushed a commit to reference the directory structure.

justaugustus commented:
I'd like to thank everyone who's reviewed thus far. It's been extremely valuable to get your input, especially from the historical aspect! :heart: 

Given that this KEP is currently marked as `provisional`, I'd like to keep in the spirit of [merging early](https://github.com/kubernetes/community/blob/master/keps/0000-kep-template.md) and iterating as we move through the KEP statuses:

> Avoid getting hung up on specific details and instead aim to get the goal of the KEP merged quickly. The best way to do this is to just start with the "Overview" sections and fill out details incrementally in follow on PRs. View anything marked as a provisional as a working document and subject to change. Aim for single topic PRs to keep discussions focused. If you disagree with what is already in a document, open a new PR with suggested changes. 

If you feel that what we've laid out is a good start and moving in the right direction, please drop a `/lgtm` on this when you have a chance. Thanks!

erictune commented:
@jdumars wrote:
> the highest stakeholders which are subproject participants first and foremost, SIGs second, consumers like the release team third, and causal observers dead last. 

I don't agree that casual observers (_users_ if you will) are dead last.

@calebamiles wrote:
>  I do like that for [smaller] projects like Rust you can see all the content in one place 
 the internet (IETF) is a bigger project, and yet still manages to have a single page view of all proposals.

jbeda commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,217 @@
+- Prevent new KEPs and design proposals from landing in [k/community]
+- Remove `kind/kep` from [k/community] once KEP migration is complete
+- Correlate existing Feature tracking issues with links to KEPs
+- Fix [KEP numbering races] by using the GitHub issue number of the KEP tracking issue
```
I'm not a huge fan of this but I guess it is ok.  We'll have sparse numbering as there will be other PRs.

One thing that I want to do regardless is encourage people to check in early.  If the latency between PR for new kep and commit is small these races will be minimal.  We can and should build some presubmit to prevent races too.

In any case, if we use the PR number as the KEP number there will still have to be a dance where people will create a change, submit a PR, update their files and submit/push another change to that PR.  That dance isn't obvious and will make this a bit harder.

jbeda commented on keps/0001a-meta-kep-implementation.md:
```diff
@@ -0,0 +1,217 @@
+
+- Create tooling to:
+  - Generate KEP directories and associated metadata
+  - Present KEPs, through some easy to use mechanism e.g., https://enhancements.k8s.io. This would be a redesigned version of https://contributor.kubernetes.io/keps/. We envision this site / repo having at least three directories:
```
We are already building out contributors.k8s.io and we should put the keps there.  That was one of the main reasons for that site.  /cc @castrojo

castrojo replied:
We can totally do this, we just need the repo URL when it's created and we can implement it. Doing it with three sections wouldn't be a problem. Hugo can be sensitive to the metadata though, it'd be great if the tooling could lint it before hugo consumes it.

jbeda commented:
@erictune:
> I dislike that https://contributor.kubernetes.io/keps/ is organized alphabetically

I agree -- the number should out in front. I think this is just a point in time issue.  @castrojo  -- Where should Eric file an issue.

jbeda commented:
Also -- merge early! Let me know when you want an LGTM label and I'll happily put it on there.  Don't want to do it now to prevent a surprise merge.

jdumars commented:
@erictune sorry I wasn't clear enough on my comment. I meant casual observers as people who do not have any stake in either consuming or generating the information. Many people "weigh in" on things as an academic exercise. I'd prefer that this site be focused on meeting user and generator needs.

castrojo commented:
@erictune and others, https://github.com/kubernetes-sigs/contributor-site/issues is where site issues should be. I am filing other issues for the design proposals and architecture designs as well. Please feel free to file issues and recommendations. Happy to share OWNERship with anyone who wants to rev faster/contribute.

justaugustus commented:
@jbeda -- thanks for the feedback. This is ready for an `/lgtm` and merge!
/hold cancel

jbeda commented:
/approve

jbeda commented:
/lgtm

justaugustus commented:
/remove-sig pm
/sig architecture
/area enhancements
